bpo-38056: overhaul Error Handlers section in codecs documentation (#15732)
python/cpython: Doc/library/codecs.rst

This module defines base classes for standard Python codecs (encoders and
decoders) and provides access to the internal Python codec registry, which
manages the codec and error handling lookup process. Most standard codecs
are :term:`text encodings <text encoding>`, which encode text to bytes (and

> reverse), but there are also codecs provided that encode text to text, and

Reviewer: ```suggestion
the reverse), but there are also codecs provided that encode text to text, and
```

bytes to bytes. Custom codecs may encode and decode between arbitrary types,

> but some module features are restricted to use specifically with

Reviewer: ```suggestion
but some module features are restricted to be used specifically with
```

> :term:`text encodings <text encoding>`, or with codecs that encode to

Reviewer: ```suggestion
 :term:`text encodings <text encoding>` or with codecs that encode to
```

This module defines base classes for standard Python codecs (encoders and
decoders) and provides access to the internal Python codec registry, which
manages the codec and error handling lookup process. Most standard codecs
are :term:`text encodings <text encoding>`, which encode text to bytes (and

> the reverse), but there are also codecs provided that encode text to text, and

Reviewer 2: (and the reverse) => and decode bytes to text

bytes to bytes. Custom codecs may encode and decode between arbitrary types,
but some module features are restricted to be used specifically with

>  :term:`text encodings <text encoding>` or with codecs that encode to

Reviewer: It looks like the indentation on this line is causing the Travis doc build issues:
```
/home/travis/build/python/cpython/Doc/library/codecs.rst:30:Unexpected indentation.
```

```suggestion
:term:`text encodings <text encoding>` or with codecs that encode to
```

Error Handlers
^^^^^^^^^^^^^^

> To simplify and standardize error handling, codecs may implement different

Reviewer 3: Please refrain to rewrap paragraphs that don't need to. Can you put those few lines like they were before?

codecs may implement different error handling schemes by
accepting the *errors* string argument.  The following string values are
defined and implemented by all standard Python codecs:
To simplify and standardize error handling, codecs may implement different
error handling schemes by accepting the *errors* string argument.

The following error handlers can be used by both ``encode`` and ``decode``
codecs:

.. tabularcolumns:: |l|L|

+-------------------------+-----------------------------------------------+
| Value                   | Meaning                                       |
+=========================+===============================================+
| ``'strict'``            | Raise :exc:`UnicodeError` (or a subclass);    |
| ``'strict'``            | Raise :exc:`UnicodeError` (or a subclass),    |
|                         | this is the default.  Implemented in          |
|                         | :func:`strict_errors`.                        |
+-------------------------+-----------------------------------------------+
| ``'ignore'``            | Ignore the malformed data and continue        |

> |                         | without further notice.  Implemented in       |

Reviewer 3: Why rewraping this one? Without an important reason, please revert it back to avoid useless diff in history and PRs.

Error Handlers
^^^^^^^^^^^^^^

To simplify and standardize error handling, codecs may implement different
error handling schemes by accepting the *errors* string argument:

      >>> 'German ß, ♬'.encode(encoding='ascii', errors='backslashreplace')
      b'German \\xdf, \\u266c'
      >>> 'German ß, ♬'.encode(encoding='ascii', errors='xmlcharrefreplace')

>       b'German &#223;, &#9836;'

Reviewer 2: I like the idea of examples. Would you mind to add an example for all available error handlers?

It may be interesting to add an example for surrogatepass which is an uncommon case.

Reviewer: @vstinner 
> I like the idea of examples. Would you mind to add an example for all available error handlers?

I'm in agreement with adding an example for anything commonly utilized, but I don't think we should necessarily add one for all of the error handlers. 

> It may be interesting to add an example for surrogatepass which is an uncommon case.

IMO, we should try to focus on having examples for the common cases. Code examples can be very helpful, but in excess they can become distracting to readers.

Reviewer 4: I like the idea of example errors as well. I wonder if the table is making this more cluttered. 

Perhaps something like:

```
Error Name
    Definition
    Example
```
would be more helpful.

Alternatively, a blank line between REPL examples would increase readability.

Reviewer: @animalize 
> If an error handleris is very easy to understand, maybe no need to give an example.

I think the simple ones could still benefit from an example, just to show the basics of how it works. Even if it's fairly simple, it may not be quite as easy to understand for someone reading over the codecs documentation for the first time.

> Or if an error handler is uncommon, maybe the reader doesn't need to be disturbed by an example, we can just use the text to describe it clearly.
IMHO these are not very common in real code: namereplace/surrogateescape/surrogatepass.

> Then only two remaining. (backslashreplace/xmlcharrefreplace)
These two examples also teach some Unicode knowledges imperceptibly: the value of code point, ß is not an ASCII character.

:+1:

.. index::
   pair: strict; error handler's name
   pair: ignore; error handler's name
   pair: replace; error handler's name
   pair: backslashreplace; error handler's name
   pair: surrogateescape; error handler's name
   single: ? (question mark); replacement character
   single: \ (backslash); escape sequence
   single: \x; escape sequence
   single: \u; escape sequence
   single: \U; escape sequence

> The following error handlers can be used with all :ref:`standard-encodings`

Reviewer: Recommended change:
```
The following error handlers can be used with all codecs that are considered
:ref:`standard-encodings`:
```

| ``'xmlcharrefreplace'`` | Replace with XML/HTML numeric character       |
|                         | reference, which is a decimal form of Unicode |
|                         | code point with format ``&#num;`` Implemented |
|                         | in :func:`xmlcharrefreplace_errors`.          |
+-------------------------+-----------------------------------------------+
| ``'namereplace'``       | Replace with ``\N{...}`` escape sequences,    |
|                         | what appears in the braces is the Name        |
|                         | property from Unicode Character Database.     |
|                         | Implemented in :func:`namereplace_errors`.    |
+-------------------------+-----------------------------------------------+

.. index::
   pair: surrogatepass; error handler's name

In addition, the following error handler is specific to the given codecs:

+-------------------+------------------------+-------------------------------------------+
| Value             | Codecs                 | Meaning                                   |
+===================+========================+===========================================+

> |``'surrogatepass'``| utf-8, utf-16, utf-32, | Allow encoding and decoding Surrogate code|
> |                   | utf-16-be, utf-16-le,  | point (``U+D800`` - ``U+DFFF``) as normal |
> |                   | utf-32-be, utf-32-le   | code point. Otherwise these codecs treat  |
> |                   |                        | the presence of Surrogate code point in   |

Reviewer 5: ```suggestion
|``'surrogatepass'``| utf-8, utf-16, utf-32, | Allow encoding and decoding surrogate code|
|                   | utf-16-be, utf-16-le,  | point (``U+D800`` - ``U+DFFF``) as normal |
|                   | utf-32-be, utf-32-le   | code point. Otherwise these codecs treat  |
|                   |                        | the presence of surrogate code point in   |
```

https://unicode.org/faq/utf_bom.html#:~:text=A%3A%20Surrogates%20are%20code%20points,DC0016%20to%20DFFF16. doesn't capitalize it.

.. function:: strict_errors(exception)

   Implements the ``'strict'`` error handling.

   Each encoding or decoding error raises a :exc:`UnicodeError`.

.. function:: ignore_errors(exception)

   Implements the ``'ignore'`` error handling.

>    Malformed data is ignored and encoding or decoding is continued without

Reviewer: ```suggestion
   Malformed data is ignored; encoding or decoding is continued without
```

   further notice.

.. function:: replace_errors(exception)

   Implements the ``'replace'`` error handling.

>    Substitutes ``?`` (ASCII character) for encoding errors, or ``U+FFFD`` (the

Reviewer: ```suggestion
   Substitutes ``?`` (ASCII character) for encoding errors or ``U+FFFD`` (the
```

   official REPLACEMENT CHARACTER) for decoding errors.

.. function:: backslashreplace_errors(exception)

   Implements the ``'backslashreplace'`` error handling.

   Malformed data is replaced by a backslashed escape sequence.

>    On encoding, use hexadecimal form of Unicode code point with formats

Reviewer: ```suggestion
   On encoding, use the hexadecimal form of Unicode code point with formats
```

>    ``\xhh`` ``\uxxxx`` ``\Uxxxxxxxx``. On decoding, use hexadecimal form of

Reviewer: ```suggestion
   ``\xhh`` ``\uxxxx`` ``\Uxxxxxxxx``. On decoding, use the hexadecimal form of
```

>    byte value with format ``\xhh``.

Reviewer: ```suggestion
   byte values with format ``\xhh``.
```

Reviewer: @animalize:
> \xhh is a single byte's value, should we keep it as is?

Oh, I see. Yes, this part can be kept as is then.

   .. versionchanged:: 3.5

>       Now works with decoding and translating.

Reviewer: ```suggestion
      Works with decoding and translating.
```

.. function:: xmlcharrefreplace_errors(exception)

   Implements the ``'xmlcharrefreplace'`` error handling (for encoding within
   :term:`text encoding` only).

   The unencodable character is replaced by an appropriate XML/HTML numeric
   character reference, which is a decimal form of Unicode code point with

>    format ``&#num;``

Reviewer: ```suggestion
   format ``&#num;``.
```

Reviewer: @animalize 
> I omitted it intentionally, so that not to cause visual confusion with the previous &#num;

The visual confusion shouldn't be an issue since you used the code blocks. Could you take a look at the section with the code blocks and the period after using ``make html`` to build the docs? It may look a bit confusing in the plaintext version, but after adding the Sphinx and reST formatting it shouldn't be an issue.

Reviewer: LGTM, adding the space looks fine.

.. function:: namereplace_errors(exception)

   Implements the ``'namereplace'`` error handling (for encoding within
   :term:`text encoding` only).

>    The unencodable character is replaced by a ``\N{...}`` escape sequence,

Reviewer: ```suggestion
   The unencodable character is replaced by a ``\N{...}`` escape sequence.
```

>    what appears in the brace is the Name property from Unicode Character

Reviewer: ```suggestion
   The sequence that appears in the brace is the *name* property from Unicode Character
```

Reviewer: @animalize 
> I'm afraid the reader may confuse, what is the "sequence" refered to in this paragraph?

Hmm, I see your point, what do you think of "The sequence of characters that appear within the braces ..."? This would differentiate it from the escape sequence without being as vague as the original "what appears" .

Reviewer: Perhaps "The set of characters ..."?

Reviewer: Also note that "brace" should be "braces", since it is enclosed within a pair of braces. I did not include that in my initial suggestion.

Reviewer: @animalize I was suggesting for "The set of characters" to replace "what appears".
 
Here are all of the changes I would suggest for the second paragraph:
```
All unencodeable characters are replaced with a ``\N{..}`` escape sequence. The set of
characters that appear in the braces represent the ``Name`` property from the Unicode
Character Database. For example, the German lowercase letter ... will be converted to
the byte sequence ...
```
Note: Includes a few new suggestions and doesn't include the line wrapping. Also, I omitted a few parts using ellipses (...).

Reviewer: @animalize:
> To be honest, I want the documentation to be simple.

Is there a particular part of the recommendation that you thought was not phrased in a simple manner? If so, I can provide an alternative with only the grammar corrections. I'd be glad to further work on improving it, that was just a first idea. 

>  Maybe the tutorial part needs a more standard language, but the reference part can be simple a bit.

Based on my experience from the documentation changes and reviews I've made so far, it seems to be the other way around for the most part.

The tutorial tends to be the less formalized part and more simple, while the language reference focuses more on technical accuracy. @JulienPalard and @willingc might be able to provide more accurate information on that though (since they're both documentation experts).

function interfaces of the stateless encoder and decoder:

> .. method:: Codec.encode(input[, errors='strict'])

Reviewer 5: ```suggestion
.. method:: Codec.encode(input, errors='strict')
```

If we give the default we don't need the square brackets to say that the param is optional.

   of the output object type in this situation.

> .. method:: Codec.decode(input[, errors='strict'])

Reviewer 5: ```suggestion
.. method:: Codec.decode(input, errors='strict')
```

   object.

>    .. method:: encode(object[, final=False])

Reviewer 5: ```suggestion
   .. method:: encode(object, final=False)
```

   object.

>    .. method:: decode(object[, final=False])

Reviewer 5: ```suggestion
   .. method:: decode(object, final=False)
```

   :func:`register_error`.

>    .. method:: read([size=-1[, chars=-1, [firstline=False]]])

Reviewer 5: ```suggestion
   .. method:: read(size=-1, chars=-1, firstline=False)
```

      available on the stream, these should be read too.

>    .. method:: readline([size=None[, keepends=True]])

Reviewer 5: ```suggestion
   .. method:: readline(size=None, keepends=True)
```

      returned.

>    .. method:: readlines([sizehint=None[, keepends=True]])

Reviewer 5: ```suggestion
   .. method:: readlines(sizehint=None, keepends=True)
```
